### PR TITLE
Fix up DI for components.

### PR DESCRIPTION
**Pull request by dereuromark**
Tackles
https://discourse.cakephp.org/t/cakephp-5-dependency-injection-inject-existing-service-into-component/12721/1

### CODE REVIEW

File: tests/TestCase/Controller/ComponentRegistryTest.php [hunk 1]
```diff
@@ -122,6 +125,31 @@ public function testLoadWithContainerAutoWiring(): void
...
+        // Register service and component with explicit arguments (as user would do)
+        $service = new TestService();
+        $container->add(TestService::class, $service);
+        $container->add(ComponentRegistry::class, $components);
+        $container->add(InjectedServiceComponent::class)
+            ->addArgument(ComponentRegistry::class)
+            ->addArgument(TestService::class);
+
+        // This should work without duplicating arguments
+        $component = $components->load(InjectedServiceComponent::class);
```
Comment (ADmad) [inline tests/TestCase/Controller/ComponentRegistryTest.php]: Can you please add/update the test with non empty config and ensure it's properly passed?

File: src/Controller/ComponentRegistry.php [hunk 2]
```diff
@@ -150,22 +151,34 @@ protected function _create(object|string $class, string $alias, array $config):
...
-                        ->addArguments($args);
-                } catch (NotFoundException) {
-                    $this->container->add($class)
-                        ->addArguments($args);
+            // Try to resolve directly from container first
+            // This allows user-configured components to work without modification
+            try {
+                /** @var \Cake\Controller\Component $instance */
+                $instance = $this->container->get($class);
+            } catch (Throwable) {
```
Comment (ADmad) [inline src/Controller/ComponentRegistry.php]: Better to catch the specific exception here instead of Throwable

File: src/Controller/ComponentRegistry.php [hunk 3]
```diff
@@ -150,22 +153,41 @@ protected function _create(object|string $class, string $alias, array $config):
...
-                        ->addArguments($args);
+            // Try to resolve directly from container first
+            // This allows user-configured components to work without modification
+            try {
+                /** @var \Cake\Controller\Component $instance */
+                $instance = $this->container->get($class);
+
+                // Merge runtime config into the component
+                if ($config) {
+                    $instance->setConfig($config);
```
Comment (ADmad) [inline src/Controller/ComponentRegistry.php]: I had a feeling this was missing earlier :)

Comment (dereuromark) [reply src/Controller/ComponentRegistry.php]: Good eye :)

**Comment by ADmad** [comment]
This is quite hackish, but probably there isn't a better way to handle this.

**Comment by dereuromark** [comment]
@markstory Might have ideas. Lets see.

**Comment by der-dave** [comment]
Hello everyone!
I reported the issue in the forums, dereuromark made this PR and asked me to add some more information here instead of forums.

**The changes in _ComponentRegistry_ are working like described in documentation.** :)

dereuromark mentioned it should be simply possible to register the Component like this:
`$container->add(OauthComponent::class);`

When doing so, I get the following Exception:
> Component App\Controller\Component\OauthComponent is registered in container but cannot be resolved.

For now I am registering the Component like the documentation says:
`$container->add(OauthComponent::class)
   ->addArgument(ComponentRegistry::class)
   ->addArgument(GenericProvider::class);`
and its working.

Thank you!
--------------------
Another problem I noticed is the following (and I dont know if its because of the cahnges in _ComponentRegistry_:
If Controller and Component have same names like:
> OauthController;
OauthComponent;

I got completely weird results and errors saying:

> [Cake\ORM\Exception\MissingTableClassException] Table class for alias Oauth could not be found. in /cakephp/src/ORM/Locator/TableLocator.php on line 249
Exception Attributes: array (
0 => ‘for alias Oauth’,
)

Its weired because I am not using TableLocator anywhere in _OauthController_ or _OauthComponent_. For now I fixed it changing name of _OauthController_ to something different.

File: src/Controller/ComponentRegistry.php (hunk 3 above)
Comment (markstory) [inline src/Controller/ComponentRegistry.php]: This and shared instances are going to be unpleasant. Should we have doc comment content about this edge case?

File: src/Controller/ComponentRegistry.php [hunk 4]
```diff
@@ -150,22 +153,41 @@ protected function _create(object|string $class, string $alias, array $config):
...
-                    $this->container->extend($class)
-                        ->addArguments($args);
-                } catch (NotFoundException) {
-                    $this->container->add($class)
-                        ->addArguments($args);
+            // Try to resolve directly from container first
+            // This allows user-configured components to work without modification
+            try {
+                /** @var \Cake\Controller\Component $instance */
+                $instance = $this->container->get($class);
```
Comment (markstory) [inline src/Controller/ComponentRegistry.php]: Instead of a try catch could you use `container->has($class)` to see if the container already has the component defined?

**Comment by markstory** [comment]
CI failures seem related.
